horovod/tensorflow/__init__.py

```diff
@@ -191,12 +200,41 @@ def allreduce_grads(grads):
                              if grad is not None and isinstance(grad, tf.IndexedSlices)
                              else grad for grad in grads]
 
-                return [allreduce(grad,
-                                  device_dense=self._device_dense,
-                                  device_sparse=self._device_sparse,
-                                  compression=self._compression)
-                        if grad is not None else grad
-                        for grad in grads]
+                # Note: Grouping does not support tf.IndexedSlices due to allgather requirement.
+                # Disabling feature if any tf.IndexedSlices entries are found.
+                # Right now, group ids/sizes are set here with basic logic. Would like to
+                # discuss ideas for more robust handling of this.
+                if (self._num_groups and self._num_groups > 0 and
+                    not any(isinstance(grad, tf.IndexedSlices) for grad in grads)):
+
+                    # Register groups/get group ids
+                    # Note: Each group needs a unique name. Currently using a basic namimg scheme
+                    # with gradient tensor names and an index. The group name is used to reassign
+                    # existing group ids in modes of operation where the group registration calls
+                    # may occur multiple times in a loop on the same set of tensors.
+                    num_grads_per_group = (len(grads) + self._num_groups - 1) // self._num_groups
+                    group_ids = [register_group(num_grads_per_group,
+                                                "%s:%s:%d" % (grads[0].name, grads[-1].name, i))
+                                 for i in range(len(grads) // num_grads_per_group)]
+                    if len(grads) % num_grads_per_group != 0:
+                        group_ids.append(register_group(len(grads) % num_grads_per_group,
+                                                        "%s:%s:%d" % (grads[0].name, grads[-1].name,
+                                                        len(grads) // num_grads_per_group + 1)))
+
+                    return [allreduce(grad,
```

Comment:
Perhaps we can simplify this to remove the else:

```
return [allreduce(grad,
                             device_dense=self._device_dense,
                             device_sparse=self._device_sparse,
                             compression=self._compression,
                             group_id=group_ids[idx // num_grads_per_group] if group_ids else -1)
            if grad is not None else grad
            for idx, grad in enumerate(grads)]
```

Then we declare `group_ids = None` above outside the if statement.

Reply:
Thanks for the suggestion! Implemented.